frontend/src/hooks/refresh.ts

```diff
@@ -0,0 +1,61 @@
+import { useSelector } from "react-redux";
+import { KialiAppState } from "../store/Store";
+import { IntervalInMilliseconds, TimeInMilliseconds } from "../types/Common";
+import { useEffect, useState } from "react";
+
+let numSubscribers = 0;
+let intervalId: null | number = null;
+
+function doTick(time?: TimeInMilliseconds) {
+  const refreshTick = new CustomEvent('refreshTick', {detail: time ?? Date.now()});
+  document.dispatchEvent(refreshTick);
+}
+
+export function triggerRefresh(time?: TimeInMilliseconds) {
+  doTick(time);
+}
+
+export default function useRefreshInterval() {
+  const refreshInterval = useSelector<KialiAppState, IntervalInMilliseconds>((state) => state.userSettings.refreshInterval);
+  const [lastRefreshAt, setLastRefreshAt] = useState<TimeInMilliseconds>(Date.now());
+  const [previousRefreshAt, setPreviousRefreshAt] = useState<TimeInMilliseconds>(lastRefreshAt);
+
+  useEffect(() => {
+    function handleTick(e: CustomEventInit<TimeInMilliseconds>) {
+      setPreviousRefreshAt(lastRefreshAt);
+      setLastRefreshAt(e.detail!);
+    }
+
+    // Subscribe
+    document.addEventListener('refreshTick', handleTick);
+    numSubscribers++;
+
+    return function () {
+      // Unsubscribe;
+      document.removeEventListener('refreshTick', handleTick);
+      numSubscribers--;
+    };
+  }, [lastRefreshAt]);
```

Comment:
I'm confused as to why we're adding dom listeners here and tracking the number of subscribers?

Is the refresh interval shared across all components or does each component have its own refresh interval?

Reply:
> Is the refresh interval shared across all components or does each component have its own refresh interval?

Yes, it is shared.

Previously, the `Refresh.tsx` component owned the `interval` and propagated the refresh through Redux which other components "listened" passively. Across Kiali pages you may notice that there is always a single set of duration-refreshInterval-refreshButton components, which hopefully gives clarity that the refresh interval is shared. If that `Refresh.tsx` component is not rendered, auto-refresh does not happen and this is why the plugin is not refreshing data right now (because in Kiosk mode, toolbars aren't rendered, thus `Refresh.tsx` is also not rendered).

I wanted to keep the shared nature of the refresh interval. So, this refresh.ts hook is what I'm trying to use for it. Without redux in play to propagate refresh events (i.e. without that `setLastRefreshAt` dispatch) I couldn't think a way to listen/fire the refresh event other than using DOM listeners. Also, since this hook can be invoked by several components on a single page, I thought it was worth to track the number of subscribers to know when to start/stop the interval - somehow I wanted to have a way to stop the interval when needed; i.e. something similar to not having the interval if the `Refresh.tsx` is not rendered - it is still not the same, but perhaps somewhat close. Sadly, I found that a few components in the masthead are watching for refreshes, meaning that once you login, auto-refresh remains active all the time :-/

Reply:
So we want to share the same state across components (the refresh interval) and have multiple components subscribe to that state but we can't use Redux because the plugin does not support redux and using hooks would give each component its own copy of the state rather than having a single shared copy? Is that right? Why does the plugin not use redux? Is this a factor of using iframes and kiosk mode?

I'm not sure this is any better but another option is to set the refresh interval in the URL and have the components watch for changes to the url. iirc some components are already doing something similar.

I'm fine with using the dom listeners for now but it feels like what we want to use (and were using) was Redux for sharing state across components. Is that it an inherent limitation of the plugin that we can't use Redux or just the way it's currently implemented with iframes and kiosk mode?

Reply:
> we can't use Redux because the plugin does not support redux and using hooks would give each component its own copy of the state rather than having a single shared copy? Is that right? Why does the plugin not use redux? Is this a factor of using iframes and kiosk mode?

At the moment I can only give answers that represent just my opinion, because currently I think everything it is somewhat vague. I cannot say that we cannot use Redux. Indeed, we _can_ because it is just an `iframe` and we can do whatever we want there. However, I do think we _may need_ to reduce Redux usage if we eventually fall on the need of re-using components between Kiali and the plugin (either by creating a new repo of shared components, or by doing refactoring within the existent repos). The Redux store is what is giving high coupling across components in the front-end. I do think we can still use Redux in a native OS plugin, but we may need to find ways to reduce the coupling and for this I think we may need to "pull up" redux state to have just a few top level components connected to Redux and, then, pass state down using standard props rather than having everything connected to Redux.

That said, I don't think this "subscription" model is what motivated the usage of DOM listeners. (read the following)

> I'm not sure this is any better but another option is to set the refresh interval in the URL and have the components watch for changes to the url. iirc some components are already doing something similar.

The refresh interval is indeed already available through the URL. Indeed, this number is still in Redux. However, when I refer to the `interval` I'm talking about the javascript timer that is setup by invoking the `setInterval` function. I don't think that can go in the URL. I couldn't think on another way of managing this timer, but suggestions are very welcome as I also don't like this approach that much.

Reply:
Thanks for answering my questions. We don't need to figure all this out now in this PR I'm just trying to get more clarity on how the different state management techniques are supposed to work together.

> I do think we can still use Redux in a native OS plugin, but we may need to find ways to reduce the coupling and for this I think we may need to "pull up" redux state to have just a few top level components connected to Redux and, then, pass state down using standard props rather than having everything connected to Redux.

Yep!

> The refresh interval is indeed already available through the URL. Indeed, this number is still managed in Redux. However, when I refer to the interval I'm talking about the javascript timer that is setup by invoking the setInterval function. I don't think that can go in the URL. I couldn't think on another way of managing this timer, but suggestions are very welcome as I also don't like this approach that much.

Managing the interval could be part of the redux action I think. The timer would just start ticking when the store gets initialized. The current approach is fine for now I'd say and is working as expected. When we go to refactor more things related to state management in the future perhaps we can revisit then.